crates/biome_deserialize/src/json.rs

```diff
@@ -582,34 +99,82 @@ where
     );
     Deserialized {
         diagnostics: errors,
-        deserialized: output,
+        deserialized,
     }
 }
 
 /// Attempts to deserialize a JSON AST, given the `Output`.
-pub fn deserialize_from_json_ast<Output>(parse: &JsonRoot) -> Deserialized<Output>
-where
-    Output: Default + VisitJsonNode + JsonDeserialize,
-{
-    let mut output = Output::default();
+pub fn deserialize_from_json_ast<Output: Deserializable>(parse: &JsonRoot) -> Deserialized<Output> {
     let mut diagnostics = vec![];
-    Output::deserialize_from_ast(parse, &mut output, &mut diagnostics);
+    let deserialized = parse
+        .value()
+        .ok()
+        .and_then(|value| Output::deserialize(value, &mut diagnostics));
     Deserialized {
         diagnostics: diagnostics.into_iter().map(Error::from).collect::<Vec<_>>(),
-        deserialized: output,
+        deserialized,
     }
 }
 
-/// Attempts to deserialize a JSON AST, given the `Output`.
-pub fn deserialize_from_json_root<Output>(parse: &JsonRoot) -> Deserialized<Output>
-where
-    Output: Default + VisitJsonNode + JsonDeserialize,
-{
-    let mut output = Output::default();
-    let mut diagnostics = vec![];
-    Output::deserialize_from_ast(parse, &mut output, &mut diagnostics);
-    Deserialized {
-        diagnostics: diagnostics.into_iter().map(Error::from).collect::<Vec<_>>(),
-        deserialized: output,
+impl DeserializableValue for AnyJsonValue {
+    fn range(&self) -> biome_rowan::TextRange {
+        AstNode::range(self)
+    }
+
+    fn deserialize<V: DeserializationVisitor>(
+        self,
+        visitor: V,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<V::Output> {
+        let range = AstNode::range(&self);
```

Comment:
Unfortunately I cannot make this change, because it is ambiguous for the compiler.

Reply:
Really?? So, what's the point of the function `fn range()` here? 🤔 Sorry I need to understand 😓

Reply:
It is for retrieving the range of a desrializable value (which is under the hood an AST Node). Thus, we have two methods named `range`: one is from the `AstNode` trait, while the other is from `DeserializableValue`. `DeserializableValue::range` is calling `AstNode::range`.

Using `DeserializableValue` instead of `AstNode<L>` improves API ergonomics (no more generics to specify) and hide low-level details.

An alternative could be of introducing a trait `Ranged` that exposes a method `range`. `AstNode` and `DeserializableValue` could require the implementation of this trait. This could avoid the duplication.

Reply:
I understand. It was a weird thing to see, for sure 🤔